### DISCUSSION TITLE
Minestom server crashes on start: OutOfMemoryError

### DISCUSSION THREAD

**Discussion by ArvedEh** [Minestom/Minestom#877] [Help] [answered]
Hello,
my Minestom server runs perfectly fine on my local machine. However, on my server, I get this error when starting the server:

```
[0,309s][warning][os,thread] Failed to start thread - pthread_create failed (EAGAIN) for attributes: stacksize: 1024k, guardsize: 0k, detached.
Error occurred during initialization of VM
java.lang.OutOfMemoryError: unable to create native thread: possibly out of memory or process/resource limits reached
```

My server is running with 4GB of RAM, about 1GB is currently used. My start command is `java -jar PlaceStom-v1.0-all.jar`,
and I also tried `java -Xmx2G -jar PlaceStom-v1.0-all.jar` and `java -Xmx1G -jar PlaceStom-v1.0-all.jar`

Looking online, I found that there can be user and system limits. Here is the result of me checking for these limits:
```
minecraft@***:~/PlaceServer$ cat /proc/sys/kernel/threads-max
3090192
minecraft@***:~/PlaceServer$ cat /etc/security/limits.conf
(Large file, but all lines are commented out)
minecraft@***:~/PlaceServer$ cat /proc/meminfo
MemTotal:        4194304 kB
MemFree:         2023444 kB
MemAvailable:    3078044 kB
Cached:          1048168 kB
```

I also get this error when trying to run `java -version`, although that worked previously (Fyi: My java version is 17, and I am running Termurin) and I get it on root as well (although I would not run the server on root, even if that was the only user it worked for)

Are there any more troubleshooting steps I can take? Should I contact my server hoster support with this?

**Comment by Eoghanmc22** [comment, marked as answer]
I am not the most familiar with this but there is something wrong with your java install or user limits (the logs you posted look like system wide limits?). check for native memory limits (jdk), check out the ulimit command, check if other processes are spawing a bunch of threads, and maybe try reinstalling the jdk. easiest option would prob be to contact ur hosting provider tho.

**Reply by ArvedEh** [reply]
Hi,
I tried `ulimit`:
```
minecraft@***:~/server$ ulimit
unlimited
```
As for the java install, I installed java 17 just for Minestom, but the existing java install (java 11) has the same errors when starting jdk 11 jars like a paper jar that used to work half a year ago

About other processes spawning threads: I doubt they reach 3090192 threads, but I have checked and there are not that many running.
I also checked the CPU usage, and I am currently using around 0.3 - 0,5 % of the CPU (the only other services are a few webservers that get like one request a day and a few discord bots)

Thanks for trying to help though, and I will contact my server hoster now.

**Reply by Eoghanmc22** [reply]
there are several flags to the `ulimit` command that could be helpful notably -a (all limits) -H (hard limit) -S (soft limit) you can find more documentations and information online although from my expereance ulimit is a bit of a painful command lol

**Reply by ArvedEh** [reply]
I talked a lot to my server hoster in the last few days and it turns out that they limit the amount of threads to 400. As the Minestom server is (by far) not the only thing running on that machine, this limit is easily reached and makes the server not viable for Minestom projects.
